ms_active_directory/core/ad_session.py

```diff
@@ -1737,11 +1753,17 @@ def find_groups_for_entities(self, entities: List[Union[str, ADObject]], attribu
             # cast all members to lowercase for a case-insensitive membership check. our normalization
             # function gave use lowercase DNs
             member_set = set(member.lower() for member in result.get(ldap_constants.AD_ATTRIBUTE_MEMBER))
+            group_sid = result.get(ldap_constants.AD_ATTRIBUTE_OBJECT_SID) if include_primary else None
             for entity_dn in entity_dns_to_entities:
-                if entity_dn in member_set:
+                if include_primary and primary_group_dict[entity_dn] == group_sid:
+                    # if result group is primary group for entity, insert it at the beginning of the list
+                    entity = entity_dns_to_entities[entity_dn]
+                    mapping_dict[entity].insert(0, result)
```

Comment:
this is a good behavior. I specifically think this is really good because, for posix compliance, many systems will just treat the "first" group membership returned by AD as the primary group if there is no primary group. so this is adaptable to those models too

Reply:
Exactly. I should probably describe this behavior in the docstring now that I think about it.